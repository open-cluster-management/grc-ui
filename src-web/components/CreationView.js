/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ******************************************************************************
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { hideResourceToolbar } from '../../lib/client/resource-helper'
import { TemplateEditor } from './TemplateEditor'
import policyTemplate from './templates/policy-template.hbs'
import Choices from './templates'
import msgs from '../../nls/platform.properties'
import _ from 'lodash'

// where to put Create/Cancel buttons
const Portals = Object.freeze({
  editBtn: 'edit-button-portal-id',
  createBtn: 'create-button-portal-id',
  cancelBtn: 'cancel-button-portal-id',
})

// controlData is converted to templateData when tempplate handlebar is rendered
//  id: becomes template variable;
//  active: becomes value replacement
//  type: what type of control to prompt user
//  available: available choices
//  reverse: the yaml object to control path--if user edits yaml, what control should be updated
//  mustExist: when validating the template for creation, this value must exist
const controlData = [
  {
    name: 'creation.view.policy.name',
    id: 'name',
    type: 'text',
    active: 'policy-grc',
    reverse: 'Policy[0].metadata.name',
    mustExist: true,
  },
  {
    name: 'creation.view.policy.namespace',
    id: 'namespace',
    type: 'singleselect',
    description: 'policy.create.namespace.tooltip',
    available: [],
    reverse: 'Policy[0].metadata.namespace',
    mustExist: true,
  },
  {
    name: 'creation.view.policy.specs',
    description: 'policy.create.specs.tooltip',
    placeholder: 'creation.view.policy.select.specs',
    id: 'specs',
    type: 'multiselect',
    available: [],
    isOneSelection: true, // close dropdown on one selection--otherwise dropdown stays open
    updateNamePrefix: 'policy-', // if user hasn't typed in a custom name, update name using this selection
    reverse: ['Policy[0].spec',],
    mustExist: true,
  },
  {
    name: 'creation.view.policy.binding',
    description: 'policy.create.selectors.tooltip',
    placeholder: 'creation.view.policy.select.selectors',
    id: 'clusters',
    type: 'multiselect',
    available: [],
    reverse: 'PlacementRule[0].spec.clusterSelector', // automatically does matchLabels && matchExpressions
    mustExist: false,
  },
  {
    name: 'creation.view.policy.standards',
    description: 'policy.create.standards.tooltip',
    placeholder: 'creation.view.policy.select.standards',
    id: 'standards',
    type: 'multiselect',
    available: ['NIST', 'NIST-CSF', 'PCI', 'FISMA', 'HIPAA'],
    reverse: 'Policy[0].metadata.annotations["policy.open-cluster-management.io/standards"]',
    cacheUserValueKey: 'create.policy.standards',
  },
  {
    name: 'creation.view.policy.categories',
    description: 'policy.create.categories.tooltip',
    placeholder: 'creation.view.policy.select.categories',
    id: 'categories',
    type: 'multiselect',
    available: ['PR.PT Protective Technology','PR.DS Data Security', 'PR.AC Identity Management Authentication and Access Control', 'PR.IP Information Protection Processes and Procedures', 'DE.CM Security Continuous Monitoring'],
    reverse: 'Policy[0].metadata.annotations["policy.open-cluster-management.io/categories"]',
    cacheUserValueKey: 'create.policy.categories',
  },
  {
    name: 'creation.view.policy.controls',
    description: 'policy.create.controls.tooltip',
    placeholder: 'creation.view.policy.select.controls',
    id: 'controls',
    type: 'multiselect',
    available: ['PR.PT-1 Audit Logging','PR.PT-3 Least Functionality','PR.DS-2 Data-in-transit','PR.DS-2 Data-at-rest','PR.AC-4 Access Control', 'PR.AC-5 Network Integrity', 'PR.IP-1 Baseline Configuration', 'DE.CM-7 Monitoring for unauthorized activity','DE.CM-8 Vulnerability scans'],
    reverse: 'Policy[0].metadata.annotations["policy.open-cluster-management.io/controls"]',
    cacheUserValueKey: 'create.policy.controls',
  },
  {
    name: 'creation.view.policy.enforce',
    description: 'policy.create.enforce.tooltip',
    id: 'enforce',
    type: 'checkbox',
    active: 'inform',
    available: ['inform', 'enforce'],  // in template, 'inform'===checkbox unchecked
    reverse: 'Policy[0].spec.remediationAction',
    mustExist: true,
  },
  {
    name: 'creation.view.policy.disabled',
    description: 'policy.create.disabled.tooltip',
    id: 'disabled',
    type: 'checkbox',
    active: 'false',
    available: ['false', 'true'],
    reverse: 'Policy[0].spec.disabled',
  },
]


export default class CreationView extends React.Component {

  static propTypes = {
    createControl: PropTypes.shape({
      createResource: PropTypes.func,
      cancelCreate: PropTypes.func,
      creationStatus: PropTypes.string,
      creationMsg: PropTypes.string
    }),
    discovered: PropTypes.object,
    fetchControl: PropTypes.shape({
      isLoaded: PropTypes.bool,
      isFailed: PropTypes.bool,
    }),
  }

  render() {
    hideResourceToolbar()
    const { locale } = this.context
    const {fetchControl, createControl, discovered} = this.props
    return (
      <TemplateEditor
        template={policyTemplate}
        controlData={getControlData(discovered, locale)}
        portals={Portals}
        fetchControl={fetchControl}
        createControl={createControl}
        type={'policy'}
        locale={locale}
      />
    )
  }
}

const getControlData = (discovered, locale) => {
  if (discovered) {
    const mergedData = _.cloneDeep(controlData)

    // add preset spec choices from yaml
    const controlMap = _.keyBy(mergedData, 'id')
    Object.values(Choices).forEach(choice=>{
      const available = _.get(controlMap, `${choice.multiselect}.available`)
      if (available) {
        available.push(choice)
      }
    })

    // add discovered choices from server
    //  add available annotations to categories, etc controls
    //  add existing policy names to name control
    const {policyNames, namespaces, annotations, clusterLabels} = discovered
    const {name, namespace, clusters, standards, categories, controls } = _.keyBy(mergedData, 'id')
    name.existing = policyNames
    namespace.available = namespaces
    clusters.available = clusterLabels
    standards.available = [...new Set([...standards.available, ...annotations.standards])]
    categories.available = [...new Set([...categories.available, ...annotations.categories])]
    controls.available = [...new Set([...controls.available, ...annotations.controls])]

    // convert message keys
    mergedData.forEach(control=>{
      ['name', 'description', 'placeholder'].forEach(key=>{
        if (control[key]) {
          control[key] = msgs.get(control[key], locale)
        }
      })
    })

    return mergedData
  }
  return controlData
}
