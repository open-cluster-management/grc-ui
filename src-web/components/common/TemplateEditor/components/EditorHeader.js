/*******************************************************************************
 * Licensed Materials - Property of IBM
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import '../scss/editor-header.scss'
import msgs from '../../../../nls/platform.properties'

class EditorHeader extends React.Component {

  static propTypes = {
    children: PropTypes.node,
    locale: PropTypes.string,
  }

  constructor (props) {
    super(props)
  }

  render() {
    const { children, locale } = this.props
    const editorToolbarTitle = msgs.get('editor.toolbar', locale)
    return (
      <div className='creation-view-yaml-header' >
        <div className='creation-view-yaml-header-toolbar' role='region' aria-label={editorToolbarTitle} id={editorToolbarTitle}>
          {children}
        </div>
      </div>
    )
  }

}

export default EditorHeader
