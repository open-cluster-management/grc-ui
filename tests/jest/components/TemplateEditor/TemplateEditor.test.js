/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import {TemplateEditor} from '../../../../src-web/components/TemplateEditor'
import renderer from 'react-test-renderer'
import policyTemplate from '../../../../src-web/components/templates/policy-template.hbs'
import { shallow } from 'enzyme'
//import { existing } from '../ComponentsTestingData'


const Portals = Object.freeze({
  createBtn: 'create-button-portal-id',
  cancelBtn: 'cancel-button-portal-id',
})

const controlData = [
  {
    name: 'creation.view.policy.name',
    id: 'name',
    type: 'text',
    active: 'policy-grc',
    reverse: 'Policy[0].metadata.name',
    mustExist: true,
  },
  {
    id: 'namespace',
    type: 'hidden',   // don't prompt for namespace--use configuration
    active: 'mcm',
    reverse: 'Policy[0].metadata.namespace',
    mustExist: true,
  },
  {
    name: 'creation.view.policy.specs',
    description: 'policy.create.specs.tooltip',
    prompt: 'creation.view.policy.select.specs',
    id: 'specs',
    type: 'multiselect',
    available: [],
    isOneSelection: true, // close dropdown on one selection--otherwise dropdown stays open
    updateNamePrefix: 'policy-', // if user hasn't typed in a custom name, update name using this selection
    reverse: [
      'Policy[0].spec.role-templates',
      'Policy[0].spec.object-templates',
      'Policy[0].spec.policy-templates'
    ],
    mustExist: true,
  },
  {
    name: 'creation.view.policy.binding',
    description: 'policy.create.selectors.tooltip',
    prompt: 'creation.view.policy.select.selectors',
    id: 'clusters',
    type: 'multiselect',
    available: [],
    reverse: 'PlacementPolicy[0].spec.clusterLabels', // automatically does matchLabels && matchExpressions
    mustExist: true,
  },
  {
    name: 'creation.view.policy.standards',
    description: 'policy.create.standards.tooltip',
    prompt: 'creation.view.policy.select.standards',
    id: 'standards',
    type: 'multiselect',
    available: ['NIST', 'PCI', 'FISMA', 'HIPAA'],
    reverse: 'Policy[0].metadata.annotations["policy.mcm.ibm.com/standards"]'
  },
  {
    name: 'creation.view.policy.categories',
    description: 'policy.create.categories.tooltip',
    prompt: 'creation.view.policy.select.categories',
    id: 'categories',
    type: 'multiselect',
    available: ['SystemAndCommunicationsProtections','SystemAndInformationIntegrity'],
    reverse: 'Policy[0].metadata.annotations["policy.mcm.ibm.com/categories"]'
  },
  {
    name: 'creation.view.policy.controls',
    description: 'policy.create.controls.tooltip',
    prompt: 'creation.view.policy.select.controls',
    id: 'controls',
    type: 'multiselect',
    available: ['MutationAdvisor','VulnerbilityAdvisor','SecretEncryption'],
    reverse: 'Policy[0].metadata.annotations["policy.mcm.ibm.com/controls"]'
  },
  {
    name: 'creation.view.policy.enforce',
    description: 'policy.create.enforce.tooltip',
    id: 'enforce',
    type: 'checkbox',
    active: 'inform',
    available: ['inform', 'enforce'],  // in template, 'inform'===checkbox unchecked
    reverse: 'Policy[0].spec.remediationAction',
    mustExist: true,
  },
]


describe('TemplateEditor component', () => {
  it('renders as expected', () => {
    const component = renderer.create(
      <TemplateEditor
        template={policyTemplate}
        controlData={controlData}
        portals={Portals}
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('on control change function', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <TemplateEditor
        template={policyTemplate}
        controlData={controlData}
        portals={Portals}
      />
    )
    const evt = {
      target: {
        value: 'value-testing'
      },
      selectedItems: ['selectedItems-testing-1', 'selectedItems-testing-2'],
    }
    expect(wrapper.instance().onChange('name', evt)).toEqual('name')
    expect(wrapper.instance().onChange('standards', evt)).toEqual('standards')
    expect(wrapper.instance().onChange('categories', evt)).toEqual('categories')
    expect(wrapper.instance().onChange('controls', evt)).toEqual('controls')
    expect(wrapper.instance().onChange('clusters', evt)).toEqual('clusters')
    //expect(wrapper.instance().onChange('enforce', evt)).toEqual('enforce')
    // expect(wrapper.instance().onChange('specs', evt)).toEqual('specs')
  })
})

describe('on editor change function', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <TemplateEditor
        template={policyTemplate}
        controlData={controlData}
        portals={Portals}
      />
    )
    expect(wrapper.instance().handleParse()).toMatchSnapshot()
  })
})

describe('handleEditorCommand function', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <TemplateEditor
        template={policyTemplate}
        controlData={controlData}
        portals={Portals}
      />
    )
    expect(wrapper.instance().handleEditorCommand('next')).toEqual('next')
    expect(wrapper.instance().handleEditorCommand('undo')).toEqual('undo')
    expect(wrapper.instance().handleEditorCommand('redo')).toEqual('redo')
    expect(wrapper.instance().handleEditorCommand('restore')).toEqual('restore')
  })
})


describe('getResourceJSON function', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <TemplateEditor
        template={policyTemplate}
        controlData={controlData}
        portals={Portals}
      />
    )
    expect(wrapper.instance().getResourceJSON()).toEqual(null)
  })
})
