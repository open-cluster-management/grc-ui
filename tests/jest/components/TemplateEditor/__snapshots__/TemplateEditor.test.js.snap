// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateEditor component renders as expected 1`] = `
<div
  className="creation-view"
>
  <div
    className="creation-view-controls-container"
  >
    <div
      className="creation-view-controls-note"
    >
      All fields marked with an asterisk (*) are mandatory.
    </div>
    <div
      className="creation-view-controls"
    >
      <div
        className="creation-view-controls-textbox"
      >
        <div
          className="creation-view-controls-textbox-title"
        >
          creation.view.policy.name
          <div
            className="creation-view-controls-must-exist"
          >
            *
          </div>
        </div>
        <div
          className="bx--form-item"
        >
          <input
            className="bx--text-input bx--text__input"
            disabled={false}
            id="name"
            onChange={[Function]}
            onClick={[Function]}
            placeholder={undefined}
            type="text"
            value="policy-grc"
          />
        </div>
      </div>
      <div
        className="creation-view-controls-multiselect"
      >
        <div
          className="creation-view-controls-multiselect-title"
        >
          creation.view.policy.specs
          <div
            className="creation-view-controls-must-exist"
          >
            *
          </div>
          <div
            className="bx--tooltip-icon"
          >
            <button
              aria-label="policy.create.specs.tooltip"
              className="bx--tooltip__trigger bx--tooltip--icon__top"
            >
              <svg
                className="info-icon"
              >
                <use
                  href="#diagramIcons_info"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          aria-invalid={undefined}
          aria-label="Choose an item"
          className="bx--multi-select bx--combo-box bx--list-box"
          data-invalid={undefined}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="listbox"
          tabIndex="-1"
        >
          <div
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            className="bx--list-box__field"
            data-toggle={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              autoComplete="off"
              className="bx--text-input"
              disabled={false}
              id="downshift-0-input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              role="combobox"
              value=""
            />
            
            <div
              className="bx--list-box__menu-icon"
            >
              <svg
                alt="Open menu"
                aria-label="Open menu"
                className={undefined}
                fill={undefined}
                fillRule="evenodd"
                focusable="false"
                height="5"
                name={undefined}
                role="img"
                style={undefined}
                viewBox="0 0 10 5"
                width="10"
              >
                <title>
                  Open menu
                </title>
                <path
                  d="M0 0l5 4.998L10 0z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="creation-view-controls-multiselect"
      >
        <div
          className="creation-view-controls-multiselect-title"
        >
          creation.view.policy.binding
          <div
            className="creation-view-controls-must-exist"
          >
            *
          </div>
          <div
            className="bx--tooltip-icon"
          >
            <button
              aria-label="policy.create.selectors.tooltip"
              className="bx--tooltip__trigger bx--tooltip--icon__top"
            >
              <svg
                className="info-icon"
              >
                <use
                  href="#diagramIcons_info"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          aria-invalid={undefined}
          aria-label="Choose an item"
          className="bx--multi-select bx--combo-box bx--list-box"
          data-invalid={undefined}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="listbox"
          tabIndex="-1"
        >
          <div
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            className="bx--list-box__field"
            data-toggle={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              autoComplete="off"
              className="bx--text-input"
              disabled={false}
              id="downshift-1-input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              role="combobox"
              value=""
            />
            
            <div
              className="bx--list-box__menu-icon"
            >
              <svg
                alt="Open menu"
                aria-label="Open menu"
                className={undefined}
                fill={undefined}
                fillRule="evenodd"
                focusable="false"
                height="5"
                name={undefined}
                role="img"
                style={undefined}
                viewBox="0 0 10 5"
                width="10"
              >
                <title>
                  Open menu
                </title>
                <path
                  d="M0 0l5 4.998L10 0z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="creation-view-controls-multiselect"
      >
        <div
          className="creation-view-controls-multiselect-title"
        >
          creation.view.policy.standards
          <div
            className="bx--tooltip-icon"
          >
            <button
              aria-label="policy.create.standards.tooltip"
              className="bx--tooltip__trigger bx--tooltip--icon__top"
            >
              <svg
                className="info-icon"
              >
                <use
                  href="#diagramIcons_info"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          aria-invalid={undefined}
          aria-label="Choose an item"
          className="bx--multi-select bx--combo-box bx--list-box"
          data-invalid={undefined}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="listbox"
          tabIndex="-1"
        >
          <div
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            className="bx--list-box__field"
            data-toggle={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              autoComplete="off"
              className="bx--text-input"
              disabled={false}
              id="downshift-2-input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              role="combobox"
              value=""
            />
            
            <div
              className="bx--list-box__menu-icon"
            >
              <svg
                alt="Open menu"
                aria-label="Open menu"
                className={undefined}
                fill={undefined}
                fillRule="evenodd"
                focusable="false"
                height="5"
                name={undefined}
                role="img"
                style={undefined}
                viewBox="0 0 10 5"
                width="10"
              >
                <title>
                  Open menu
                </title>
                <path
                  d="M0 0l5 4.998L10 0z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="creation-view-controls-multiselect"
      >
        <div
          className="creation-view-controls-multiselect-title"
        >
          creation.view.policy.categories
          <div
            className="bx--tooltip-icon"
          >
            <button
              aria-label="policy.create.categories.tooltip"
              className="bx--tooltip__trigger bx--tooltip--icon__top"
            >
              <svg
                className="info-icon"
              >
                <use
                  href="#diagramIcons_info"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          aria-invalid={undefined}
          aria-label="Choose an item"
          className="bx--multi-select bx--combo-box bx--list-box"
          data-invalid={undefined}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="listbox"
          tabIndex="-1"
        >
          <div
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            className="bx--list-box__field"
            data-toggle={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              autoComplete="off"
              className="bx--text-input"
              disabled={false}
              id="downshift-3-input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              role="combobox"
              value=""
            />
            
            <div
              className="bx--list-box__menu-icon"
            >
              <svg
                alt="Open menu"
                aria-label="Open menu"
                className={undefined}
                fill={undefined}
                fillRule="evenodd"
                focusable="false"
                height="5"
                name={undefined}
                role="img"
                style={undefined}
                viewBox="0 0 10 5"
                width="10"
              >
                <title>
                  Open menu
                </title>
                <path
                  d="M0 0l5 4.998L10 0z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="creation-view-controls-multiselect"
      >
        <div
          className="creation-view-controls-multiselect-title"
        >
          creation.view.policy.controls
          <div
            className="bx--tooltip-icon"
          >
            <button
              aria-label="policy.create.controls.tooltip"
              className="bx--tooltip__trigger bx--tooltip--icon__top"
            >
              <svg
                className="info-icon"
              >
                <use
                  href="#diagramIcons_info"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          aria-invalid={undefined}
          aria-label="Choose an item"
          className="bx--multi-select bx--combo-box bx--list-box"
          data-invalid={undefined}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="listbox"
          tabIndex="-1"
        >
          <div
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            className="bx--list-box__field"
            data-toggle={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              autoComplete="off"
              className="bx--text-input"
              disabled={false}
              id="downshift-4-input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              role="combobox"
              value=""
            />
            
            <div
              className="bx--list-box__menu-icon"
            >
              <svg
                alt="Open menu"
                aria-label="Open menu"
                className={undefined}
                fill={undefined}
                fillRule="evenodd"
                focusable="false"
                height="5"
                name={undefined}
                role="img"
                style={undefined}
                viewBox="0 0 10 5"
                width="10"
              >
                <title>
                  Open menu
                </title>
                <path
                  d="M0 0l5 4.998L10 0z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="creation-view-controls-checkbox"
      >
        <div
          className="bx--form-item bx--checkbox-wrapper"
        >
          <input
            checked={false}
            className="bx--checkbox"
            id="enforce"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="bx--checkbox-label checkbox"
            htmlFor="enforce"
            title={null}
          >
            <span
              className="bx--visually-hidden"
            >
              
            </span>
          </label>
        </div>
        <div>
          creation.view.policy.enforce
        </div>
        <div
          className="bx--tooltip-icon"
        >
          <button
            aria-label="policy.create.enforce.tooltip"
            className="bx--tooltip__trigger bx--tooltip--icon__top"
          >
            <svg
              className="info-icon"
            >
              <use
                href="#diagramIcons_info"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="creation-view-yaml"
  >
    <div
      className="creation-view-yaml-header"
    >
      <div
        className="creation-view-yaml-header-title"
      >
        Policy YAML
      </div>
      <div
        aria-label={undefined}
        className="creation-view-yaml-header-toolbar"
        id={undefined}
        role="region"
      >
        <div
          className="editor-bar"
        >
          <div
            aria-label="Search"
            className="editor-bar-search"
            id="Search"
            role="region"
          >
            <div
              aria-labelledby="search-name-label"
              className="bx--search bx--search--sm"
              role="search"
            >
              <svg
                alt=""
                aria-label=""
                className="bx--search-magnifier"
                fill={undefined}
                fillRule="evenodd"
                height="16"
                name={undefined}
                role="img"
                style={undefined}
                viewBox="0 0 16 16"
                width="16"
              >
                <title>
                  
                </title>
                <path
                  d="M6.5 12a5.5 5.5 0 1 0 0-11 5.5 5.5 0 0 0 0 11zm4.936-1.27l4.563 4.557-.707.708-4.563-4.558a6.5 6.5 0 1 1 .707-.707z"
                />
              </svg>
              <label
                className="bx--label"
                htmlFor="search-name"
                id="search-name-label"
              >
                
              </label>
              <input
                aria-label="Search"
                className="bx--search-input"
                id="search-name"
                onChange={[Function]}
                placeholder="Search"
                type="text"
              />
              <button
                aria-label={undefined}
                className="bx--search-close bx--search-close--hidden"
                onClick={[Function]}
                type="button"
              >
                <svg
                  alt="Provide a description that will be used as the title"
                  aria-label="Provide a description that will be used as the title"
                  className={undefined}
                  fill={undefined}
                  fillRule="evenodd"
                  height="16"
                  name={undefined}
                  role="img"
                  style={undefined}
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <title>
                    Provide a description that will be used as the title
                  </title>
                  <path
                    d="M8 6.586L5.879 4.464 4.464 5.88 6.586 8l-2.122 2.121 1.415 1.415L8 9.414l2.121 2.122 1.415-1.415L9.414 8l2.122-2.121-1.415-1.415L8 6.586zM8 16A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            aria-label="Previous"
            className="editor-bar-button  disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Previous"
          >
            <svg
              height="1rem"
              width="1rem"
            >
              <use
                href="#diagramIcons_previous"
              />
            </svg>
          </div>
          <div
            aria-label="Next"
            className="editor-bar-button  disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Next"
          >
            <svg
              height="1rem"
              width="1rem"
            >
              <use
                href="#diagramIcons_next"
              />
            </svg>
          </div>
          <div
            className="editor-bar-spacer"
          />
          <div
            aria-label="Undo"
            className="editor-bar-button  disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Undo"
          >
            <svg
              height="1rem"
              width="1rem"
            >
              <use
                href="#diagramIcons_undo"
              />
            </svg>
          </div>
          <div
            aria-label="Redo"
            className="editor-bar-button  disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Redo"
          >
            <svg
              height="1rem"
              width="1rem"
            >
              <use
                href="#diagramIcons_redo"
              />
            </svg>
          </div>
          <div
            aria-label="Restore"
            className="editor-bar-button  disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Restore"
          >
            <svg
              height="1rem"
              width="1rem"
            >
              <use
                href="#diagramIcons_restore"
              />
            </svg>
          </div>
          <div
            className="editor-bar-spacer"
          />
        </div>
      </div>
    </div>
    
    <div
      className="yamlEditorContainer"
    >
      <div
        id="brace-editor"
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`on editor change function renders as expected 1`] = `
"apiVersion: policy.mcm.ibm.com/v1alpha1
kind: Policy
metadata:
  name: policy-grc
  namespace: mcm
  annotations:
    policy.mcm.ibm.com/standards: 
    policy.mcm.ibm.com/categories: 
    policy.mcm.ibm.com/controls: 
spec:
  complianceType: musthave
  remediationAction: inform
  disabled: false
  namespaces:
    exclude: [\\"kube-*\\"]
    include: [\\"default\\"]
---
apiVersion: mcm.ibm.com/v1alpha1
kind: PlacementBinding
metadata:
  name: binding-policy-grc
  namespace: mcm
placementRef:
  name: placement-policy-grc
  kind: PlacementRule
  apiGroup: apps.open-cluster-management.io
subjects:
- name: policy-grc
  kind: Policy
  apiGroup: policy.mcm.ibm.com
---
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
  name: placement-policy-grc
  namespace: mcm
spec:
  clusterSelector:
    matchExpressions:
      []
"
`;
